### DISCUSSION TITLE
Use compiler macros to distinguish architectures?

### DISCUSSION THREAD

**Discussion by MinetaS** [official-stockfish/Stockfish#5527] [General]
Instead of setting every flag for each architecture extension, we can use macros defined by compiler.

GCC x86 for example: 
* Defined flags: https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/i386-c.cc
* x86 options: https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html

This can remove most parts in current Makefile, instead users can set their CPU type (or native).
`make build ARCH=native` to `g++ -march=native`, or `make build ARCH=skylake` to `g++ -march=skylake`

For example, if arch is set to skylake, the compiler sets the following macros:
`__MMX__`, `__SSE__`, `__SSE2__`, `__SSE3__`, `__SSSE3__`, `__SSE4_1__`, `__SSE4_2__`, `__POPCNT__`, `__AVX__`, `__AVX2__`, `__BMI__`, `__BMI2__`, `__PRFCHW__` (and others)

To add any other architecture, it's not required to modify Makefile and implement the feature detection.

**Comment by Disservin** [comment]
since when are these available and will clang/gcc cover all of those? 
how will we specify a different arch even if that arch isnt support by the current hw ? i.e. for ci builds

**Reply by MinetaS** [reply]
I'm pretty sure compilers that meet the minimum requirement of fishtest have this. But what do you mean by "cover" - if extension is not supported by a compiler it's already impossible to generate such instructions.

To specify the different architecture you can just set `-march=<cpu>` as we currently do like `ARCH=...`, it's irrelevant to which extension a hardware supports. (It's like cross compile, just the binary won't execute normally on it)

**Reply by MinetaS** [reply]
To give an exact answer of the first question, GCC at least has supported since the very old version.
This is the oldest online doc I could find, see Intel 386 Options. https://gcc.gnu.org/onlinedocs/gcc-2.95.3/gcc_2.html#SEC16

**Reply by Disservin** [reply]
ah right so instead of specifying an arch you specify the family name, sorry missed that :D 
feel free to open a rfc pr, im curious in how much this would actually cleanup

this would also help if we ever decide to use a different build system

**Comment by vondele** [comment]
This is something to consider (early in the next release cycle). 

Things that come to my mind:
* how do we handle the release binaries, they are more important than native compiles? 
* can we make sure we test all available targets ?

**Reply by MinetaS** [reply]
I've been testing a few versions and came to the conclusion that it's better to keep the current system using build profiles, notably due to the reasons you mentioned (difficult to decide which targets to release/maintain). Yet I believe it would be still convenient to use compiler defined macros, which means:

1. Replace all `USE_XXX` macros with corresponding ones (`__XXX__`)
2. Remove all `-DUSE_XXX` CXXFLAGS in Makefile

... assuming we keep only GNU compatible compilers in the supported compiler list.

Meanwhile making an extra option for end user to specify compiler-type architecture could be an alternative (i.e. `CXXARCH=skylake`), while keep everything as it is.

**Reply by mstembera** [reply]
I know we no longer support MSVC officially but SF has always compiled under MSVC as it still does today.  It's not the MSVC compiled binaries anyone cares about but the development environment plus debugging integration.  I do all my dev work under MSVC and would be effectively locked out from future contributions to SF.  All I ask is that whatever change may happen doesn't break the ability to compile under MSVC.  Note the Makefile is not used by MSVC at all so anything that happens there is irrelevant.

**Reply by MinetaS** [reply]
@mstembera Could you let me know what compiler command line you use?

**Reply by MinetaS** [reply]
I assume you have a set of `/D USE_XXX` if you don't use Makefile. If that's the case replacing `/D USE_XXX` with `/D __XXX__` should be good.

**Reply by mstembera** [reply]
Yes that makes sense.  My command line exported from the gui is:
/JMC /permissive- /MP /ifcOutput "x64\Debug\" /GS /W3 /Zc:wchar_t /ZI /Gm- /Od /sdl /Fd"x64\Debug\vc143.pdb" /Zc:inline /fp:precise /D "_MBCS" /D "USE_SSE2" /D "USE_SSE3" /D "USE_SSSE3" /D "USE_POPCNT" /D "USE_SSE41" /D "USE_PEXT" /D "USE_AVX2" /D "USE_AVX512" /errorReport:prompt /WX- /Zc:forScope /RTC1 /Gd /MDd /std:c++20 /FC /Fa"x64\Debug\" /EHsc /nologo /Fo"x64\Debug\" /Fp"x64\Debug\SF.pch" /diagnostics:column
